server/src/main/java/org/opensearch/cluster/routing/allocation/decider/AwarenessAllocationDecider.java

```diff
@@ -211,47 +215,94 @@ private Decision underCapacity(ShardRouting shardRouting, RoutingNode node, Rout
         return allocation.decision(Decision.YES, NAME, "node meets all awareness attribute requirements");
     }
 
+    private static Set<String> getAttributeValues(ShardRouting shardRouting, RoutingAllocation allocation, String awarenessAttribute) {
+        if (shardRouting.isSearchOnly()) {
+            return allocation.routingNodes().searchNodesPerAttributesCounts(awarenessAttribute);
+        }
+        return allocation.routingNodes().nodesPerAttributesCounts(awarenessAttribute);
+    }
+
     private int getCurrentNodeCountForAttribute(
         ShardRouting shardRouting,
         RoutingNode node,
         RoutingAllocation allocation,
         boolean moveToNode,
         String awarenessAttribute
     ) {
-        // build the count of shards per attribute value
-        final String shardAttributeForNode = getAttributeValueForNode(node, awarenessAttribute);
-        int currentNodeCount = 0;
-        final List<ShardRouting> assignedShards = allocation.routingNodes().assignedShards(shardRouting.shardId());
+        // Get the attribute value for the current node
+        String shardAttributeForNode = getAttributeValueForNode(node, awarenessAttribute);
+
+        List<ShardRouting> assignedShards = getAssignedShards(allocation, shardRouting);
+
+        // Count assigned shards with matching attribute values
+        int currentNodeCount = countMatchingShards(assignedShards, allocation, awarenessAttribute, shardAttributeForNode);
```

Comment:
`countMatchingShards` and `adjustNodeCount` together looks logically equivalent to the existing logic in `getCurrentNodeCountForAttribute` ? I'd prefer we didn't introduce the diff here.

The only diff i see is getting SR specific routings instead of 
```
        final List<ShardRouting> assignedShards = allocation.routingNodes().assignedShards(shardRouting.shardId());
```